# Segmenting and Clustering Neighbourhoods in Toronto

This is the peer graded assignment on the clustering of the neighbourhoods in Toronto. 

I have included all three questions in this workbook and have labelled the questions as 1, 2 and 3. 


### Import Libraries

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
from geopy.geocoders import Nominatim
import matplotlib.pyplot as plt

## Question 1 Scraping the Wikipedia page to get the postal codes

The first question asks to scrape wikipedia and create the dataframe in the form posted. 

In [2]:
# Loading raw data, all the tables in the wikipedia page...
raw_data = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M",flavor='bs4')

In [3]:
# Extracting the data I need and creating the dataframe
postal_codes = raw_data[0]
postal_codes = postal_codes[['Postal Code','Borough','Neighborhood']]

#### Inspecting the dataframe

In [4]:
postal_codes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Postal Code   180 non-null    object
 1   Borough       180 non-null    object
 2   Neighborhood  180 non-null    object
dtypes: object(3)
memory usage: 4.3+ KB


In [5]:
postal_codes.describe(include='all')

Postal Code       Borough  Neighborhood
count          180           180           180
unique         180            11           100
top            M8K  Not assigned  Not assigned
freq             1            77            77

*From the info above, we can see that there are no duplicate Postal Code values and there are 77 instances of "Not assigned". Therefore we can remove the "Not assigned" and do not need any further processing on the "Postal Code" column*

In [6]:
postal_codes_final = postal_codes[postal_codes['Borough'] != 'Not assigned'].reset_index(drop=True)

In [7]:
postal_codes_final.describe(include='all')

Postal Code     Borough Neighborhood
count          103         103          103
unique         103          10           99
top            M4T  North York    Downsview
freq             1          24            4

In [8]:
#Checking for null values
postal_codes_final.isnull().sum()

Postal Code     0
Borough         0
Neighborhood    0
dtype: int64

*No null values in the final dataframe*

In [9]:
postal_codes_final.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [10]:
postal_codes_final.shape

(103, 3)

**Number of Rows of the dataframe is 103**

****

## Question 2: Adding the Latitude and Longitude data to the dataframe

Question 2 asks you to get the latitude and longitude data for the neighbourhoods from the previous question and append it to the existing dataframe.

****Unfortunately the geocoder library was not working. I decided below to use the csv file provided in the assignment.****

In [11]:
# Load Coordinates CSV file
coord_df = pd.read_csv('http://cocl.us/Geospatial_data')

In [12]:
df_final = pd.merge(postal_codes_final,coord_df,on ='Postal Code')

In [13]:
df_final

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

***

## Question 3: Exploring and clustering the neighbourhoods

As per the question, I decided to focus on only the boroughs that contain the word '**Toronto**'. So the first step was to filter the dataframe for entries containing **Toronto**.

In [14]:
# Filter out all neighbourhoods that contain the name "Toronto"
df_toronto = df_final[df_final['Borough'].str.contains('Toronto')].reset_index(drop=True)

In [15]:
df_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

**Location data analysis begins**

In [16]:
import json
from geopy.geocoders import Nominatim
from pandas.io.json import json_normalize
import folium

In [17]:
# Get coordinates of Toronto

address = 'Toronto, Canada'
geolocator = Nominatim(user_agent = "toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print("The geographical coordinates of Toronto are:{},{}".format( latitude, longitude))

The geographical coordinates of Toronto are:43.6534817,-79.3839347


In [18]:
#Create Toronto Map
toronto_map = folium.Map(location=[latitude,longitude], zoom_start= 13)

In [19]:
### Adding postal codes to the map...

for latitude , longitude, postal_code, neighborhood in zip(df_toronto['Latitude'],df_toronto['Longitude'],df_toronto['Postal Code'],df_toronto['Neighborhood']):
    label='{},{}'.format(postal_code,neighborhood)
    label = folium.Popup(label,parse_html=True)
    folium.CircleMarker(
    [latitude,longitude],
    radius = 5,
    popup=label,
    fill=True,
    fill_color='blue',
    fill_opacity=0.7,
    parse_html=False).add_to(toronto_map)

**Draw Map with postal codes and neighbourhoods added**

In [20]:
# Display map showing the various postal codes and corresponding neighbourhoods
toronto_map

**Foursquare calls**

In [21]:
### Defining Foursquare Credentials and conducting analyses

import requests

CLIENT_ID = 'QMAOL0EJQ0I0TLP4EK50DS4WEX5SUX5I4MDYO00V1GO3JORY'
CLIENT_SECRET= '53CQLIUKBESM4JF52KHHDEVCRKFWDH0Y44YS5MDWRG0ZPLBV'
VERSION = '20180605'
LIMIT=100

In [22]:
# Extract coordinates of the first postal code
n_lat = df_toronto.loc[0,'Latitude']
n_long = df_toronto.loc[0,'Longitude']

In [23]:
# Create the first URL to foursquare using the first postal code coordinates. This will allow us to inspect the returned 
# json file to understand how the data is structured. This will help later to extract the needed data.
radius = 500
limit = 100


url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID,
    CLIENT_SECRET,
    VERSION,
    n_lat,
    n_long,
    radius,
    limit)



#### Initial Foursquare data evaluation

In [24]:
# Get results from foursquare API

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eef7079bae9a2001bd1c506'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 45,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

In [25]:
# Filter results to show venues
venues = results['response']['groups'][0]['items']
venues

[{'reasons': {'count': 0,
   'items': [{'summary': 'This spot is popular',
     'type': 'general',
     'reasonName': 'globalInteractionReason'}]},
  'venue': {'id': '54ea41ad498e9a11e9e13308',
   'name': 'Roselle Desserts',
   'location': {'address': '362 King St E',
    'crossStreet': 'Trinity St',
    'lat': 43.653446723052674,
    'lng': -79.3620167174383,
    'labeledLatLngs': [{'label': 'display',
      'lat': 43.653446723052674,
      'lng': -79.3620167174383}],
    'distance': 143,
    'postalCode': 'M5A 1K9',
    'cc': 'CA',
    'city': 'Toronto',
    'state': 'ON',
    'country': 'Canada',
    'formattedAddress': ['362 King St E (Trinity St)',
     'Toronto ON M5A 1K9',
     'Canada']},
   'categories': [{'id': '4bf58dd8d48988d16a941735',
     'name': 'Bakery',
     'pluralName': 'Bakeries',
     'shortName': 'Bakery',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_',
      'suffix': '.png'},
     'primary': True}],
   'photos': {'count': 0, 'grou

In [26]:
# Convert json file into pandas dataframe

explored_venues = json_normalize(venues)
explored_venues.head()

C:\Users\User\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


referralId  reasons.count  \
0  e-0-54ea41ad498e9a11e9e13308-0              0   
1  e-0-53b8466a498e83df908c3f21-1              0   
2  e-0-574c229e498ebb5c6b257902-2              0   
3  e-0-50760559e4b0e8c7babe2497-3              0   
4  e-0-566e1294498e3f6629006bc3-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id                venue.name venue.location.address  \
0  54ea41ad498e9a11e9e13308          Roselle Desserts          362 King St E   
1  53b8466a498e83df908c3f21             Tandem Coffee          368 King St E   
2  574c229e498ebb5c6b257902    Cooper Koo Family YMCA          461 Cherry St   
3  50760559e4b0e8c7babe2497       Body Blitz Spa East   497 King Street East   
4  566e1294498e3f6629006bc3  Dominion Pub and Kitchen  500 Queen Street East   

        venue.location.crossStreet  venue.location.lat  venue.location.lng  \
0                       Trinity St           43.653447          -79.362017   
1                    at Trinity St           43.653559          -79.361809   
2                              NaN           43.653249          -79.358008   
3  btwn Sackville St and Sumach St           43.654735          -79.359874   
4                              NaN           43.656919          -79.358967   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': 43.65344672305267...                      143   
1  [{'label': 'display', 'lat': 43.65355870959944...                      122   
2  [{'label': 'display', 'lat': 43.65324910177244...                      239   
3  [{'label': 'display', 'lat': 43.65473505045365...                       80   
4  [{'label': 'display', 'lat': 43.65691857501867...                      325   

  venue.location.postalCode venue.location.cc venue.location.city  \
0                   M5A 1K9                CA             Toronto   
1                       NaN                CA             Toronto   
2                   M5A 0H7                CA             Toronto   
3                   M5A 1L9                CA             Toronto   
4                   M5A 1T9                CA             Toronto   

  venue.location.state venue.location.country  \
0                   ON                 Canada   
1                   ON                 Canada   
2                   ON                 Canada   
3                   ON                 Canada   
4                   ON                 Canada   

                     venue.location.formattedAddress  \
0  [362 King St E (Trinity St), Toronto ON M5A 1K...   
1  [368 King St E (at Trinity St), Toronto ON, Ca...   
2        [461 Cherry St, Toronto ON M5A 0H7, Canada]   
3  [497 King Street East (btwn Sackville St and S...   
4  [500 Queen Street East, Toronto ON M5A 1T9, Ca...   

                                    venue.categories  venue.photos.count  \
0  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...                   0   
1  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...                   0   
2  [{'id': '52e81612bcbc57f1066b7a37', 'name': 'D...                   0   
3  [{'id': '4bf58dd8d48988d1ed941735', 'name': 'S...                   0   
4  [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...                   0   

  venue.photos.groups venue.venuePage.id venue.location.neighborhood  
0                  []                NaN                         NaN  
1                  []                NaN                         NaN  
2                  []                NaN                         NaN  
3                  []                NaN                         NaN  
4                  []                NaN                         NaN

In [27]:
filtered_columns = ['venue.name','venue.categories','venue.location.lat','venue.location.lng']

In [28]:
explored_venues = explored_venues.loc[:,filtered_columns]

In [29]:
explored_venues.head()

venue.name  \
0          Roselle Desserts   
1             Tandem Coffee   
2    Cooper Koo Family YMCA   
3       Body Blitz Spa East   
4  Dominion Pub and Kitchen   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...           43.653447   
1  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...           43.653559   
2  [{'id': '52e81612bcbc57f1066b7a37', 'name': 'D...           43.653249   
3  [{'id': '4bf58dd8d48988d1ed941735', 'name': 'S...           43.654735   
4  [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...           43.656919   

   venue.location.lng  
0          -79.362017  
1          -79.361809  
2          -79.358008  
3          -79.359874  
4          -79.358967

#### Function used in the lab was used to extract the categories of the various venues

In [30]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [31]:
explored_venues['venue.categories'] = explored_venues.apply(get_category_type, axis=1)


In [32]:
explored_venues.rename(columns={'venue.name':'Name','venue.categories':'Categories','venue.location.lat':'Latitude','venue.location.lng':'Longitude'},inplace=True)

#### Final dataframe of the selected Postal Code. This process was however used to extract one set of venues data. It will be automated for all the Postal Codes in the next section.

In [33]:
explored_venues.head()

Name           Categories   Latitude  Longitude
0          Roselle Desserts               Bakery  43.653447 -79.362017
1             Tandem Coffee          Coffee Shop  43.653559 -79.361809
2    Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3       Body Blitz Spa East                  Spa  43.654735 -79.359874
4  Dominion Pub and Kitchen                  Pub  43.656919 -79.358967

****After the analysis above, I now have a process to call, process and clean up the json data into a pandas dataframe.
The next step is to automate the process for all the postal codes in our area of concern. I re-used the code from the lab to get the nearby venues and clean up the data.****

In [34]:
### Function from Lab that gets the venues for all the neighborhoods

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [35]:
###Get all the data for the Toronto neighborhoods...

toronto_venues = getNearbyVenues(names=df_toronto['Postal Code'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

M5A
M7A
M5B
M5C
M4E
M5E
M5G
M6G
M5H
M6H
M5J
M6J
M4K
M5K
M6K
M4L
M5L
M4M
M4N
M5N
M4P
M5P
M6P
M4R
M5R
M6R
M4S
M5S
M6S
M4T
M5T
M4V
M5V
M4W
M5W
M4X
M5X
M4Y
M7Y


In [36]:
# Explore the dataframe
toronto_venues.shape
toronto_venues.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1623 entries, 0 to 1622
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Postal Code             1623 non-null   object 
 1   Neighborhood Latitude   1623 non-null   float64
 2   Neighborhood Longitude  1623 non-null   float64
 3   Venue                   1623 non-null   object 
 4   Venue Latitude          1623 non-null   float64
 5   Venue Longitude         1623 non-null   float64
 6   Venue Category          1623 non-null   object 
dtypes: float64(4), object(3)
memory usage: 88.9+ KB


In [37]:
toronto_venues.head()

Postal Code  Neighborhood Latitude  Neighborhood Longitude  \
0         M5A               43.65426              -79.360636   
1         M5A               43.65426              -79.360636   
2         M5A               43.65426              -79.360636   
3         M5A               43.65426              -79.360636   
4         M5A               43.65426              -79.360636   

                      Venue  Venue Latitude  Venue Longitude  \
0          Roselle Desserts       43.653447       -79.362017   
1             Tandem Coffee       43.653559       -79.361809   
2    Cooper Koo Family YMCA       43.653249       -79.358008   
3       Body Blitz Spa East       43.654735       -79.359874   
4  Dominion Pub and Kitchen       43.656919       -79.358967   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                  Pub

In [38]:
### How many venues were returned per neighborhood

toronto_venues.groupby('Postal Code').count()


Neighborhood Latitude  Neighborhood Longitude  Venue  \
Postal Code                                                         
M4E                              4                       4      4   
M4K                             43                      43     43   
M4L                             20                      20     20   
M4M                             40                      40     40   
M4N                              3                       3      3   
M4P                              8                       8      8   
M4R                             21                      21     21   
M4S                             35                      35     35   
M4T                              2                       2      2   
M4V                             17                      17     17   
M4W                              4                       4      4   
M4X                             48                      48     48   
M4Y                             76                      76     76   
M5A                             45                      45     45   
M5B                            100                     100    100   
M5C                             79                      79     79   
M5E                             58                      58     58   
M5G                             65                      65     65   
M5H                             94                      94     94   
M5J                            100                     100    100   
M5K                            100                     100    100   
M5L                            100                     100    100   
M5N                              3                       3      3   
M5P                              6                       6      6   
M5R                             20                      20     20   
M5S                             34                      34     34   
M5T                             63                      63     63   
M5V                             14                      14     14   
M5W                             97                      97     97   
M5X                            100                     100    100   
M6G                             17                      17     17   
M6H                             17                      17     17   
M6J                             44                      44     44   
M6K                             22                      22     22   
M6P                             25                      25     25   
M6R                             15                      15     15   
M6S                             34                      34     34   
M7A                             32                      32     32   
M7Y                             18                      18     18   

             Venue Latitude  Venue Longitude  Venue Category  
Postal Code                                                   
M4E                       4                4               4  
M4K                      43               43              43  
M4L                      20               20              20  
M4M                      40               40              40  
M4N                       3                3               3  
M4P                       8                8               8  
M4R                      21               21              21  
M4S                      35               35              35  
M4T                       2                2               2  
M4V                      17               17              17  
M4W                       4                4               4  
M4X                      48               48              48  
M4Y                      76               76              76  
M5A                      45               45              45  
M5B                     100              100             100  
M5C                      79               79              79  
M5E                      58               58     

In [39]:
### How many unique categories

print('There are {} unique venue categories'.format(len(toronto_venues['Venue Category'].unique())))

There are 236 unique venue categories


### Analyse each neighbourhood

#### Since the venue categories are categorical data, I need to onehot encode them before further analysis

In [40]:
toronto_venues_onehot = pd.get_dummies(toronto_venues,columns=['Venue Category'],prefix = "", prefix_sep='')

In [41]:
# Size of dataframe
print((toronto_venues.shape))

(1623, 7)


In [42]:
toronto_venues_onehot.head()

Postal Code  Neighborhood Latitude  Neighborhood Longitude  \
0         M5A               43.65426              -79.360636   
1         M5A               43.65426              -79.360636   
2         M5A               43.65426              -79.360636   
3         M5A               43.65426              -79.360636   
4         M5A               43.65426              -79.360636   

                      Venue  Venue Latitude  Venue Longitude  \
0          Roselle Desserts       43.653447       -79.362017   
1             Tandem Coffee       43.653559       -79.361809   
2    Cooper Koo Family YMCA       43.653249       -79.358008   
3       Body Blitz Spa East       43.654735       -79.359874   
4  Dominion Pub and Kitchen       43.656919       -79.358967   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Bu

In [43]:
toronto_onehot = toronto_venues_onehot.drop(['Neighborhood Latitude', 'Neighborhood Longitude','Venue', 'Venue Latitude', 'Venue Longitude'],axis=1)

In [44]:
toronto_onehot.shape

(1623, 237)

In [45]:
toronto_grouped = toronto_onehot.groupby('Postal Code').mean().reset_index()

In [46]:
toronto_grouped.head()

Postal Code  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0         M4E                0.0      0.0                 0.0           0.0   
1         M4K                0.0      0.0                 0.0           0.0   
2         M4L                0.0      0.0                 0.0           0.0   
3         M4M                0.0      0.0                 0.0           0.0   
4         M4N                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.023256   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.050000   
4             0.0              0.0               0.0             0.000000   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0               0.0                 0.0            0.0        0.0         0.0   
1               0.0                 0.0            0.0        0.0         0.0   
2               0.0                 0.0            0.0        0.0         0.0   
3               0.0                 0.0            0.0        0.0         0.0   
4               0.0                 0.0            0.0        0.0         0.0   

   Bagel Shop    Bakery   Bank    Bar  Baseball Stadium  Basketball Stadium  \
0         0.0  0.000000  0.000  0.000               0.0                 0.0   
1         0.0  0.023256  0.000  0.000               0.0                 0.0   
2         0.0  0.000000  0.000  0.000               0.0                 0.0   
3         0.0  0.050000  0.025  0.025               0.0                 0.0   
4         0.0  0.000000  0.000  0.000               0.0                 0.0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0    0.0              0.0       0.0         0.0                 0.0     0.0   
1    0.0              0.0       0.0         0.0                 0.0     0.0   
2    0.0              0.0       0.0         0.0                 0.0     0.0   
3    0.0              0.0       0.0         0.0                 0.0     0.0   
4    0.0              0.0       0.0         0.0                 0.0     0.0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0        0.00            0.0   0.000000       0.0                   0.0   
1        0.00            0.0   0.046512       0.0                   0.0   
2        0.05            0.0   0.000000       0.0                   0.0   
3        0.00            0.0   0.025000       0.0                   0.0   
4        0.00            0.0   0.000000       0.0                   0.0   

   Breakfast Spot   Brewery  Bubble Tea Shop  Building  Burger Joint  \
0             0.0  0.000000         0.000000       0.0           0.0   
1             0.0  0.023256         0.023256       0.0           0.0   
2             0.0  0.050000         0.000000       0.0           0.0   
3             0.0  0.050000         0.000000       0.0           0.0   
4             0.0  0.000000         0.000000       0.0           0.0   

   Burrito Place  Bus Line  Butcher      Café  Cajun / Creole Restaurant  \
0           0.00  0.000000      0.0  0.000000                        0.0   
1           0.00  0.000000      0.0  0.023256                        0.0   
2           0.05  0.000000      0.0  0.000000                   

In [47]:
toronto_grouped.shape

(39, 237)

****Next I checked for the top 5 venues in each Postal Code****

In [48]:
num_top_venues = 5

for hood in toronto_grouped['Postal Code']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Postal Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M4E----
               venue  freq
0       Neighborhood  0.25
1  Health Food Store  0.25
2              Trail  0.25
3                Pub  0.25
4  Martial Arts Dojo  0.00


----M4K----
                    venue  freq
0        Greek Restaurant  0.19
1      Italian Restaurant  0.07
2             Coffee Shop  0.07
3  Furniture / Home Store  0.05
4      Frozen Yogurt Shop  0.05


----M4L----
                  venue  freq
0        Sandwich Place  0.10
1  Fast Food Restaurant  0.10
2                   Gym  0.05
3    Italian Restaurant  0.05
4                  Park  0.05


----M4M----
                 venue  freq
0                 Café  0.10
1          Coffee Shop  0.08
2              Brewery  0.05
3  American Restaurant  0.05
4               Bakery  0.05


----M4N----
               venue  freq
0               Park  0.33
1        Swim School  0.33
2           Bus Line  0.33
3  Afghan Restaurant  0.00
4      Movie Theater  0.00


----M4P----
              venue  freq
0               Gym  0

In [49]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

***Next I look to create a dataframe showing the top 10 venues per Postal Code... I re-used the code from the lab***

In [50]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         M4E                 Trail          Neighborhood   
1         M4K      Greek Restaurant           Coffee Shop   
2         M4L        Sandwich Place  Fast Food Restaurant   
3         M4M                  Café           Coffee Shop   
4         M4N                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub     Health Food Store   Distribution Center   
1    Italian Restaurant             Bookstore    Frozen Yogurt Shop   
2                  Park     Fish & Chips Shop             Pet Store   
3             Gastropub                Bakery               Brewery   
4              Bus Line           Event Space     Electronics Store   

         6th Most Common Venue   7th Most Common Venue  \
0                 Dessert Shop      Dim Sum Restaurant   
1               Ice Cream Shop  Furniture / Home Store   
2                Movie Theater             Pizza Place   
3          American Restaurant             Yoga Studio   
4  Eastern European Restaurant     Dumpling Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Diner        Discount Store            Yoga Studio  
1              Yoga Studio                   Pub            Pizza Place  
2                      Pub            Restaurant          Burrito Place  
3  Comfort Food Restaurant    Seafood Restaurant         Sandwich Place  
4               Donut Shop      Doner Restaurant                Dog Run

In [51]:
from sklearn.cluster import KMeans
wcss = []
#toronto_grouped_clustering = toronto_grouped.drop('Postal Code', 1)
for i in range(1, 11):
    kmeans = KMeans(n_clusters = i, init = 'k-means++', random_state = 42)
    kmeans.fit(toronto_grouped_clustering)
    wcss.append(kmeans.inertia_)
plt.plot(range(1, 11), wcss)
plt.title('The Elbow Method')
plt.xlabel('Number of clusters')
plt.ylabel('WCSS')
plt.vlines(5,ymin=1,ymax=3,color='red')
plt.vlines(4,ymin=1,ymax=3,color='red')
plt.show()



NameError: name 'toronto_grouped_clustering' is not defined

#### The above analysis was to find the optimum number of clusters. The graph however was not as helpful as I would have liked. However, from the graph, you will notice that the "elbow" or the last big drop was between 4 and 5 clusters before going into a gentle decline. Therefore 5 clusters was chosen as optimum.


In [ ]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering )

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

#### Clustered Postal Codes clusters are added back into the main dataframe below.

In [ ]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Postal Code'), on='Postal Code')

toronto_merged.head() # check the last columns!

### Final mapping of the various clusters. 

In [ ]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postal Code'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

---

### Analysis of the various Clusters.

#### Cluster 0

In [ ]:
cluster1 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
%matplotlib inline

fig = plt.figure(figsize=(15,3))
ax1 = fig.add_subplot(131)
ax2 = fig.add_subplot(132)
ax3 = fig.add_subplot(133)
cluster1['1st Most Common Venue'].value_counts().plot(kind='bar',ax=ax1)
cluster1['2nd Most Common Venue'].value_counts().plot(kind='bar',ax=ax2)
cluster1['3rd Most Common Venue'].value_counts().plot(kind='bar',ax=ax3)
ax1.set_title('1st Most Common Venue')
ax2.set_title('2nd Most Common Venue')
ax3.set_title('3rd Most Common Venue')

As can be seen above, Cluster 0 is the largest cluster, with most of the neighborhoods belonging to this cluster. From the most common venues, it is clear that "Coffee Shops", "Cafes" and "Restaurants" dominate as the most common venues. This cluster can be seen as the food (coffee shop, restaurant and cafes) venues cluster.  

---

### Cluster 1

In [ ]:
cluster2 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
%matplotlib inline

fig = plt.figure(figsize=(15,3))
ax1 = fig.add_subplot(131)
ax2 = fig.add_subplot(132)
ax3 = fig.add_subplot(133)
cluster2['1st Most Common Venue'].value_counts().plot(kind='bar',ax=ax1)
cluster2['2nd Most Common Venue'].value_counts().plot(kind='bar',ax=ax2)
cluster2['3rd Most Common Venue'].value_counts().plot(kind='bar',ax=ax3)
ax1.set_title('1st Most Common Venue')
ax2.set_title('2nd Most Common Venue')
ax3.set_title('3rd Most Common Venue')

From the above, we see that cluster 1 has only one neighbourhood. "Garden" is the most common venue as well as "Music Venue". This appears to be an area of the city that is known for its natural beauty. 

---

### Cluster 2

In [ ]:
cluster3 = toronto_merged.loc[manhattan_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
%matplotlib inline

fig = plt.figure(figsize=(15,3))
ax1 = fig.add_subplot(131)
ax2 = fig.add_subplot(132)
ax3 = fig.add_subplot(133)
cluster3['1st Most Common Venue'].value_counts().plot(kind='bar',ax=ax1)
cluster3['2nd Most Common Venue'].value_counts().plot(kind='bar',ax=ax2)
cluster3['3rd Most Common Venue'].value_counts().plot(kind='bar',ax=ax3)
ax1.set_title('1st Most Common Venue')
ax2.set_title('2nd Most Common Venue')
ax3.set_title('3rd Most Common Venue')

The cluster above shows a spread of venues around the Postal code. Trail shows up as one of the most common venues as well as park. Using this in conjunction with the location of the postal code (near Cheltenham Park) explains why this postal code was not grouped with the "Food" cluster, cluster 0. 

---

### Cluster 3

In [ ]:
cluster4 = toronto_merged.loc[manhattan_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
%matplotlib inline

fig = plt.figure(figsize=(15,3))
ax1 = fig.add_subplot(131)
ax2 = fig.add_subplot(132)
ax3 = fig.add_subplot(133)
cluster4['1st Most Common Venue'].value_counts().plot(kind='bar',ax=ax1)
cluster4['2nd Most Common Venue'].value_counts().plot(kind='bar',ax=ax2)
cluster4['3rd Most Common Venue'].value_counts().plot(kind='bar',ax=ax3)
ax1.set_title('1st Most Common Venue')
ax2.set_title('2nd Most Common Venue')
ax3.set_title('3rd Most Common Venue')

Cluster 3 is the health/fitness cluster. Gym and Gym/Fitness Center appears as 2 of the 1st most common venues. Also, Trail appears in the 2nd most common venue and Gym appears once again in the 3rd most common venue

---

### Cluster 4

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

The final cluster is different from the others in that, the Jewelry store is the most common venue. 

---

## Conclusion

The small subsection of the Toronto boroughs (boroughs containing the word "Toronto") I used in my analysis, shows very little diversity. Most of the neighbourhoods were clustered into one big cluster mainly comprising of Coffee Shops and Cafes. Therefore one can conclude that the majority of the neighbourhoods analysed are similar. 